### PR TITLE
Add `Overlay.maybeOf`, make `Overlay.of` return a non-nullable instance

### PR DESCRIPTION
**Pull request by gspencergoog**
## Description

Adds `Overlay.maybeOf` and makes `Overlay.of` return a non-nullable instance. This was missed when we did our non-nullable by default migrations.

## Tests
 - Added a test of `maybeOf`, updated the `of` test.

### CODE REVIEW

**Comment by goderbauer** [comment]
Looks like the analyzer is not happy.

File: packages/flutter/lib/src/widgets/overlay.dart [hunk 1]
```diff
@@ -349,25 +355,23 @@ class Overlay extends StatefulWidget {
...
+    final OverlayState? result = maybeOf(context, rootOverlay: rootOverlay);
     assert(() {
-      if (debugRequiredFor != null && result == null) {
+      if (result == null) {
         final List<DiagnosticsNode> information = <DiagnosticsNode>[
           ErrorSummary('No Overlay widget found.'),
-          ErrorDescription('${debugRequiredFor.runtimeType} widgets require an Overlay widget ancestor for correct operation.'),
-          ErrorHint('The most common way to add an Overlay to an application is to include a MaterialApp or Navigator widget in the runApp() call.'),
-          DiagnosticsProperty<Widget>('The specific widget that failed to find an overlay was', debugRequiredFor, style: DiagnosticsTreeStyle.errorProperty),
+          ErrorDescription('${debugRequiredFor?.runtimeType ?? 'some'} widgets require an Overlay widget ancestor for correct operation.'),
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/overlay.dart:372]: Should the "some" be capitalized? Looks like its at the beginning of a sentence?

File: packages/flutter/lib/src/widgets/overlay.dart [hunk 2]
```diff
@@ -349,25 +355,23 @@ class Overlay extends StatefulWidget {
...
   ///
   /// If `rootOverlay` is set to true, the state from the furthest instance of
-  /// this class is given instead. Useful for installing overlay entries
-  /// above all subsequent instances of [Overlay].
+  /// this class is given instead. Useful for installing overlay entries above
+  /// all subsequent instances of [Overlay].
   ///
   /// This method can be expensive (it walks the element tree).
-  static OverlayState? of(
+  static OverlayState of(
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/overlay.dart:362]: Add a see also to maybeOf?

File: packages/flutter/lib/src/widgets/overlay.dart [hunk 3]
```diff
@@ -376,7 +380,30 @@ class Overlay extends StatefulWidget {
...
+  ///
+  /// ```dart
+  /// OverlayState? overlay = Overlay.maybeOf(context)!;
+  /// ```
+  ///
+  /// If `rootOverlay` is set to true, the state from the furthest instance of
+  /// this class is given instead. Useful for installing overlay entries above
+  /// all subsequent instances of [Overlay].
+  ///
+  /// This method can be expensive (it walks the element tree).
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/overlay.dart:399]: add a "see also" to of?

**Review by goderbauer — APPROVED** [review]
LGTM

File: packages/flutter/lib/src/widgets/overlay.dart (hunk 3 above)
Comment (gspencergoog) [reply packages/flutter/lib/src/widgets/overlay.dart:399]: Done.

File: packages/flutter/lib/src/widgets/overlay.dart (hunk 1 above)
Comment (gspencergoog) [reply packages/flutter/lib/src/widgets/overlay.dart:372]: Done.

File: packages/flutter/lib/src/widgets/overlay.dart (hunk 2 above)
Comment (gspencergoog) [reply packages/flutter/lib/src/widgets/overlay.dart:362]: Done.

**Pull request merged**